Formal2024/Exercises/ex5.lean

```diff
@@ -0,0 +1,82 @@
+import Library
+
+/-! # Part 1 : myEven and myOdd
+  definitions of myEven and myOdd natural numbers using `inductive` -/
+
+inductive myEven : Nat → Prop
+| zero : myEven 0
+| two_plus_n : ∀ n, myEven n → myEven (n + 2)
+
+inductive myOdd : Nat → Prop
+| one : myOdd 1
+| two_plus_n : ∀ n, myOdd n → myOdd (n + 2)
+
+/-! ## basic properties of myEven and myOdd -/
+
+theorem myEven_or_myOdd : ∀ n, myEven n ∨ myOdd n := sorry
+
+theorem myOdd_succ_myEven : ∀ n, myOdd n → myEven (n + 1) := sorry
+
+theorem myEven_succ_myOdd : ∀ n, myEven n → myOdd (n + 1) := sorry
+
+/-! ### addition -/
+
+theorem myEven_add_myOdd : ∀ m n, myEven m → myOdd n → myOdd (m + n) := sorry
+
+theorem myOdd_add_myEven : ∀ m n, myOdd m → myEven n → myOdd (m + n) := sorry
+
+theorem myOdd_add_myOdd : ∀ m n, myOdd m → myOdd n → myEven (m + n) := sorry
+
+theorem myEven_add_myEven : ∀ m n, myEven m → myEven n → myEven (m + n) := sorry
+
+/-! ### multiplication -/
+
+theorem myEven_mul : ∀ m n, myEven m → myEven (m * n) := sorry
+
+theorem mul_myEven : ∀ m n, myEven n → myEven (m * n) := sorry
+
+theorem myOdd_mul_myOdd : ∀ m n, myOdd m → myOdd n → myOdd (m * n) := sorry
+
+/-! ### power -/
+
+theorem myEven_pow : ∀ m n, myEven m → myEven (m ^ n) := sorry
+
+theorem myOdd_pow : ∀ m n, myOdd m → myOdd (m ^ n) := sorry
+
+/-! # Part 2 : uniqueness of natural numbers with `≤` -/
+
+theorem Nat.exists_unique_le : ∃! n m : Nat, m ≤ n := sorry
+
```

Comment:
Did we introduce exist only before?

Reply:
Haha - which is why this is a good exercise! At a first glance, you think this is wrong, but then you think about it and actually do it, and you see it's right!

We can introduce `∃!` next week? Or we can just briefly mention it before the exercise?

Reply:
I'd say it's ok if we just explain it here, but I'd also put a warning on it being a tricky exercise